### PR TITLE
Specify Arrow as an external dependency 

### LINKED ISSUE #213: Arrow library is not externalized
:beetle:

In 5.5.1, the arrow library is not externalized in the compiled bundle. That means that if someone wants to check whether the dataframe is an instance of an arrow dataframe, they cannot easily do it. the problem is that the constructors are not the same (the connector has their own version of arrow, not the same that someone else might use to check the type).

Because of this issue, the check in https://github.com/vega/vega-loader-arrow/blob/36e1d0d2b0050f6e207327e78367007c4d25e28b/src/arrow.js#L20 does not work.

### PR DESCRIPTION
**Pull request by jclay**
Do not pack Arrow source files into the final bundle, rely on them being provided externally. This should fix #213.

### CODE REVIEW

File: src/mapd-con-es6.js [hunk 1]
```diff
@@ -23,7 +23,7 @@ import EventEmitter from "eventemitter3"
 
 import MapDClientV2 from "./mapd-client-v2"
 import processQueryResults from "./process-query-results"
-import { Table } from "apache-arrow"
+import * as Arrow from "apache-arrow"
```
Comment (domoritz) [inline src/mapd-con-es6.js:26]: Is this change necessary?

**Comment by israelvicars** [comment]
@jclay Do we need to rebuild so that the arrow source files are excluded?

**Comment by jclay** [comment]
@israelvicars good catch. I'm wondering what the current practice is with keeping built artifacts like this in the repo. Do you see a reason for keeping them checked in going forward?

**Comment by israelvicars** [comment]
@jclay Yeah, that's been a matter of some debate. I've missed some of the previous repartee. My understanding is that pre-building is a practice with several deps to make them more accessible to a subset of users that would benefit from that. Dropping in a  That's rather vague and useless, just recalling echos of conversations... Because it's a practice across several repos, I think it merits more digging and consensus before we change practices. But we should change as quickly as possible if we can get some new consensus. I'd like to avoid doing it, personally.

**Comment by domoritz** [comment]
`5.5.2-rc3` doesn’t seem to work for me. See https://observablehq.com/d/12c6ce12edb5f725

**Comment by jclay** [comment]
@domoritz this should be resolved now. Are you able to give `5.5.2@rc4` a try? 

Observable's require was picking up the package as a `UMD` module, so when the cell was evaluated:

```
MapdCon = require("@mapd/connector@5.5.2-rc3").catch(() => window.MapdCon)
```

`catch` was never used. If we use `then` in place, it will properly override this to use the `MapdCon` defined on window.

I've now fixed some issues with our UMD packaging so that the following will work:

```
MapdCon = {
    const MapdCon = await require("@mapd/connector@5.5.2-rc4")
    return MapdCon.Connector
}
```

require of course returns a module, rather than the `MapdCon` class, so I've updated this to export `Connector` to clarify.

**Comment by domoritz** [comment]
Even better, this works now in Observable (note the even simpler import):

```
MapdCon = require("@mapd/connector@5.5.2-rc4")

connector = new MapdCon.Connector()
      .protocol("https")
      .host("metis.mapd.com")
      .port("443")
      .dbName("mapd")
      .user("mapd")
      .password("HyperInteractive")
      .connectAsync()
```

or if you want the connector directly

```
MapdCon = (await require("@mapd/connector@5.5.2-rc4")).Connector
```

I wonder, could you make `Connector` the default export so that we may not need the indirection anymore?

**Comment by domoritz** [comment]
This still does not seem to fix https://github.com/omnisci/mapd-connector/issues/213. 

Screen Shot 2021-01-06 at 10 53 34

@jheer do you know a way to make https://github.com/vega/vega-loader-arrow/blob/36e1d0d2b0050f6e207327e78367007c4d25e28b/src/arrow.js#L20 more robust to different ways the arrow table is constructed?

**Comment by jclay** [comment]
@domoritz OK, good to hear we're making some progress on this in Observable at least. I was hopeful that the external bundle would have solved #213. @israelvicars perhaps someone on the FE team would know a better way to tackle this.

**Comment by jclay** [comment]
@domoritz The only way I've found to handle export default is with the following:

Change the entrypoint export:

```js
export { MapdCon as default } from "./mapd-con-es6";
```

However, Observable's use of d3-require seems to still need the following:

```js
const MapdCon = (await require("http://localhost:8080/browser-connector.js")).default;
```

I can't determine why it doesn't seem to handle the default export, well...by default. Am I missing something?

**Comment by domoritz** [comment]
Can you do

```
import MapdCon from "./mapd-con-es6";
export = MapdCon;
```

**Comment by bmatcuk** [comment]
I think I've figured this out, but I couldn't push my changes to this PR since @jclay's PR is based on a fork - I created my own PR here: https://github.com/omnisci/mapd-connector/pull/218

**Comment by bmatcuk** [comment]
Closed in favor of #218
